### PR TITLE
Add DTLS handshake fragmentation test

### PR DESCRIPTION
**Pull request by piotr-now**
## Description
Add test for DTLS handshake fragmentation

## Status
**READY**

### CODE REVIEW

File: tests/suites/test_suite_ssl.function [hunk 1]
```diff
@@ -3680,6 +3737,44 @@ void handshake_serialization( )
...
+    {
+        TEST_ASSERT( srv_hs_fragmented >= 1 );
+    }
+    /* Test if the client received fragmented handshake */
+    if( expected_cli_hs_fragmentation )
+    {
+        TEST_ASSERT( cli_hs_fragmented >= 1 );
+    }
+    /* The goto below is used to avoid an "unused label" warning.*/
+    goto exit;
```
Comment (Patater) [inline tests/suites/test_suite_ssl.function]: This `goto` seems weird. Why do we need it here but not in other tests?

File: tests/suites/test_suite_ssl.function [hunk 2]
```diff
@@ -1651,6 +1697,9 @@ void perform_handshake( handshake_test_options* options )
         TEST_ASSERT( mbedtls_endpoint_init( &server, MBEDTLS_SSL_IS_SERVER,
                                             options->pk_alg, NULL, NULL, NULL ) == 0 );
     }
+
+    mbedtls_ssl_conf_authmode( &server.conf, options->srv_auth_mode );
```
Comment (Patater) [inline tests/suites/test_suite_ssl.function:1711]: Why was this added? What does it have to do with handshake fragmentation testing? Should this be added in its own commit?

File: tests/suites/test_suite_ssl.function (hunk 1 above)
Comment (piotr-now) [reply tests/suites/test_suite_ssl.function]: Yes, this `goto exit;` is unnecessary here because it is already in TEST_ASSERT. In the other tests it is needed to prevent warning. After Andrzej's refactoring some of tests does not have any TEST_ASSERT, but this `exit:` label is added automatically during compilation (by script). I will remove it

File: tests/suites/test_suite_ssl.function (hunk 2 above)
Comment (piotr-now) [reply tests/suites/test_suite_ssl.function:1711]: By default, server does not check client's certificates so I had to set it manually to be required. This forces the client to send more data to the server during handshake. Without that handshake messages from the client would be to small to be fragmented.

File: tests/suites/test_suite_ssl.function (hunk 1 above)
Comment (piotr-now) [reply tests/suites/test_suite_ssl.function]: Removed

File: tests/suites/test_suite_ssl.function (hunk 2 above)
Comment (Patater) [reply tests/suites/test_suite_ssl.function:1711]: OK, I see how it is related to fragmentation now. Could you please explain this aspect of your implementation in the git commit message?

File: tests/suites/test_suite_ssl.function [hunk 3]
```diff
@@ -3680,6 +3737,42 @@ void handshake_serialization( )
...
+    options.cli_log_obj = &cli_pattern;
+    options.srv_log_fun = log_analyzer;
+    options.cli_log_fun = log_analyzer;
+
+    perform_handshake( &options );
+
+    /* Test if the server received fragmented handshake */
+    if( expected_srv_hs_fragmentation )
+    {
+        TEST_ASSERT( srv_hs_fragmented >= 1 );
```
Comment (AndrzejKurek) [inline tests/suites/test_suite_ssl.function]: Is this sufficient? Why won't we check for exact values, passed as a test parameter?

File: tests/suites/test_suite_ssl.function [hunk 4]
```diff
@@ -6,13 +6,42 @@
...
+ * library. If it is set to the library with log_pattern it will count instances
+ * of the pattern in the logged messages.
+ */
+static void log_analyzer( void *ctx, int level,
+                          const char *file, int line,
+                          const char *str )
+{
+    (void)level; (void)line; (void)file;
+
+    log_pattern *p = (log_pattern*)ctx;
```
Comment (AndrzejKurek) [inline tests/suites/test_suite_ssl.function]: Space missing

File: tests/suites/test_suite_ssl.function [hunk 5]
```diff
@@ -6,13 +6,42 @@
...
+    if( NULL != p &&
+        NULL != p->pattern &&
+        NULL != p->counter &&
+        NULL != strstr( str, p->pattern ) )
+    {
+        ++(*( p->counter ) );
```
Comment (AndrzejKurek) [inline tests/suites/test_suite_ssl.function]: Space missing

File: tests/suites/test_suite_ssl.function [hunk 6]
```diff
@@ -3680,6 +3737,42 @@ void handshake_serialization( )
...
+void handshake_fragmentation( int mfl, int expected_srv_hs_fragmentation, int expected_cli_hs_fragmentation)
+{
+    handshake_test_options options;
+    log_pattern srv_pattern, cli_pattern;
+    size_t srv_hs_fragmented = 0;
+    size_t cli_hs_fragmented = 0;
+
+    srv_pattern.pattern = cli_pattern.pattern = "found fragmented DTLS handshake";
+    srv_pattern.counter = &srv_hs_fragmented;
+    cli_pattern.counter = &cli_hs_fragmented;
```
Comment (Patater) [inline tests/suites/test_suite_ssl.function]: Where is `cli_pattern.pattern` set?

File: tests/suites/test_suite_ssl.function (hunk 3 above)
Comment (piotr-now) [reply tests/suites/test_suite_ssl.function]: I was thinking about it too, but it is pretty hard to determinate how much data should be transferred during handshake and in how many fragments. I decided that the fact of any fragmentation is sufficient enough for this test.

Comment (AndrzejKurek) [reply tests/suites/test_suite_ssl.function]: What do you mean by `hard to determine`? Is it time consuming for the programmer, is it non-deterministic, or is there no way to tell the exact number of fragments? I can see that in `ssl-opt.sh` application data tests check the exact number and size of fragments, while DTLS handshake fragmentation tests only check if the handshake message was fragmented, so I'm inclined towards accepting it, but I would also like to know the background a bit more.

File: tests/suites/test_suite_ssl.function [hunk 7]
```diff
@@ -6,13 +6,42 @@
...
 
+
+typedef struct log_pattern
+{
+    const char *pattern;
+    size_t *counter;
+} log_pattern;
+
+/*
+ * This function can be added as a debug output for log messages from mbedtls
```
Comment (AndrzejKurek) [inline tests/suites/test_suite_ssl.function]: This is a bit unclear, I would rather state that: 
```suggestion
 * This function can be passed to mbedtls to receive output logs from it. In this case, it will count the instances of a log_pattern in the received logged messages.
```

File: tests/suites/test_suite_ssl.function [hunk 8]
```diff
@@ -3680,6 +3737,42 @@ void handshake_serialization( )
...
+    options.mfl = mfl;
+    options.srv_auth_mode = MBEDTLS_SSL_VERIFY_REQUIRED;
+    options.srv_log_obj = &srv_pattern;
+    options.cli_log_obj = &cli_pattern;
+    options.srv_log_fun = log_analyzer;
+    options.cli_log_fun = log_analyzer;
+
+    perform_handshake( &options );
+
+    /* Test if the server received fragmented handshake */
```
Comment (AndrzejKurek) [inline tests/suites/test_suite_ssl.function]:
```suggestion
    /* Test if the server received a fragmented handshake */
```

File: tests/suites/test_suite_ssl.function [hunk 9]
```diff
@@ -3680,6 +3737,42 @@ void handshake_serialization( )
...
+    if( expected_srv_hs_fragmentation )
+    {
+        TEST_ASSERT( srv_hs_fragmented >= 1 );
+    }
+    /* Test if the client received fragmented handshake */
```
Comment (AndrzejKurek) [inline tests/suites/test_suite_ssl.function]:
```suggestion
    /* Test if the client received a fragmented handshake */
```

**Review by AndrzejKurek — CHANGES_REQUESTED** [review]
Minor changes requested.

File: tests/suites/test_suite_ssl.function [hunk 10]
```diff
@@ -6,13 +6,42 @@
...
+/*
+ * This function can be added as a debug output for log messages from mbedtls
+ * library. If it is set to the library with log_pattern it will count instances
+ * of the pattern in the logged messages.
+ */
+static void log_analyzer( void *ctx, int level,
+                          const char *file, int line,
+                          const char *str )
+{
+    (void)level; (void)line; (void)file;
```
Comment (AndrzejKurek) [inline tests/suites/test_suite_ssl.function]: Please move those to separate lines, and add a space after `(void)`, as I'm pretty sure I haven't seen this convention used anywhere in the library.

File: tests/suites/test_suite_ssl.function (hunk 3 above)
Comment (piotr-now) [reply tests/suites/test_suite_ssl.function]: If we want to send application data, we decide how much of them should be. For handshake the message length is unknown it should be estimate. It is related with the certificates, allowed cipher suites and etc. I think it is not easy to estimate the biggest message size during handshake. I could hardcoded the current number of fragments with which this test is performed (if it is good enough)

File: tests/suites/test_suite_ssl.function (hunk 2 above)
Comment (piotr-now) [reply tests/suites/test_suite_ssl.function:1711]: I added an explanation to the commit message

**Comment by piotr-now** [comment]
I changed counter inside log_pattern to be a variable instead of pointer to variable

File: tests/suites/test_suite_ssl.function [hunk 11]
```diff
@@ -5,6 +5,35 @@
...
+static void log_analyzer( void *ctx, int level,
+                          const char *file, int line,
+                          const char *str )
+{
+    (void) level;
+    (void) line;
+    (void) file;
+
+    log_pattern *p = (log_pattern*) ctx;
+    if( NULL != p &&
```
Comment (AndrzejKurek) [inline tests/suites/test_suite_ssl.function:30]: This formatting is a bit strange (since you're well below the line limit), but it's a minor.

**Review by AndrzejKurek — APPROVED** [review]

**Review by Patater — APPROVED** [review]
LGTM
